### ISSUE TITLE
GSoC: Extension the official test suite of JSON Schema

### ISSUE THREAD

**Issue by benjagm** [json-schema-org/community#315]
**Extension the official test suite of JSON Schema**

**Brief Description**
The JSON Schema test suite is a collection of reviewed examples which support creating compliant implementations of the JSON Schema specification(s) by verifying their validation result against the expected ones. The existing test suite, which covers validation results (whether a particular piece of data is valid or invalid under a schema) is mature, containing tests across all modern JSON Schema versions. Most recently, JSON Schema has grown to include the notion of annotations, which are pieces of metadata collected whilst processing an instance using a schema. These annotations are used to define various JSON Schema keywords, but also are useful for building JSON Schema tooling beyond simple valid/invalid validation results.
An initial set of annotation tests has been [added to the suite](https://github.com/json-schema-org/JSON-Schema-Test-Suite/tree/main/output-tests) to begin work on providing tests for this functionality, but it needs extending to fully cover the set of annotation-related functionality in the specification.

**Expected Outcomes**
Outcomes of this work should include:
- Carefully read and understand the JSON Schema specifications with a focus on annotations, combing the spec for test cases which should be added, and raising discussion questions in cases where behavior seems unclear in the specification
- Production of a suite of JSON test cases which fully cover of all JSON Schema keywords which produce annotations, by adding corresponding tests for each one (or each scenario) alongside their expected annotation output for all (3) versions of JSON Schema which contain annotation-related behavior
- Authoring of a simple annotation collection library, either by re-implementing "throwaway" annotation support in an existing implementation which already has it (e.g. https://github.com/gregsdennis/json-everything) or in a net-new implementation. The purpose of doing so should specifically be to ensure that the suite of tests added "fully cover" the relevant functionality (i.e. that they test all scenarios needed for annotation collection)

**Skills Required**
- Some notion of what JSON Schema is -- or equivalently some experience using a similar data validation tool including writing SQL Schemas for databases
- Basic experience with [any of a broad set of languages with JSON Schema implementations](https://json-schema.org/implementations.html#validators) or curiosity to learn enough about them to write a simple program
- Some Python experience a strong plus

**Mentors**

@Julian 

**Expected Difficulty**
Easy-Medium

**Expected Time Commitment**
175 Hour

**Comment by gregsdennis** [comment]
> re-implementing "throwaway" annotation support

@Julian can you expand on this?

**Comment by Julian** [comment]
Basically just "write an implementation of annotation collection alongside the suite to make sure it has good test coverage", but if done for an implementation which already has annotation support then the new annotation implementation would be purely a throwaway.

**Comment by gregsdennis** [comment]
I think I'm stuck on what you mean that an annotation is "throwaway."  Can you give an example?

**Comment by Julian** [comment]
An annotation isn't throwaway some *code* is throwaway. E.g. if you were writing a validator to check if the existing suite has good coverage but with no intention of releasing the validator you wrote, it'd be throwaway code.

**Comment by gregsdennis** [comment]
So what do you mean by

> re-implementing "throwaway" annotation support in an existing implementation which already has it

If the implementation already has annotation support, why re-implement it?  Why not just use what's there?  I guess I really don't understand what you're trying to achieve with the second outcome.

**Comment by Julian** [comment]
Somehow we're misunderstanding each other I guess, I'm saying something quite simple, writing throwaway code as a way to either prototype something or verify something is a quite normal thing, no? The point is to verify that the test suite being written is good. You don't know that unless you try to use it to write an implementation, and then whilst doing so you gauge whether the tests are helping you not make mistakes. You then fill in any gaps in the tests. The point isn't to get an annotation support implementation, indeed if you want that you just use the existing one.  The point is to use the process of writing an implementation to gauge the quality of the tests (and whether the implementation ends up being usable is explicitly either not desired or not expected). I did the same exact thing when initially writing the test suite we have, and also did it again this past month while writing the referencing implementation -- I wrote a suite and an implementation, fully intending to throw the implementation away and restart it, it was just there to evaluate the suite.
